src/postprocess/chi0.jl

```diff
@@ -105,31 +105,105 @@ precondprep!(P::FunctionPreconditioner, ::Any) = P
+                            ψk_extra=nothing, εk_extra=nothing,
+                            tol_cg=1e-12, verbose=false)
     basis = Hk.basis
     kpoint = Hk.kpoint
+    model = basis.model
+    temperature = model.temperature
 
-    # we err on the side of caution here by applying Q *a lot*
-    # there are optimizations to be made here
+    # projector onto the orthogonal of occupied states
     Q(ϕ) = ϕ - ψk * (ψk' * ϕ)
-    function QHQ(ϕ)
-        Qϕ = Q(ϕ)
-        Q(Hk * Qϕ - εnk * Qϕ)
-    end
-    precon = PreconditionerTPA(basis, kpoint)
-    precondprep!(precon, ψnk)
-    function f_ldiv!(x, y)
-        x .= Q(precon \ Q(y))
+
+    if iszero(temperature) && !isnothing(ψk_extra)
+        # We use a Schur decomposition of the orthogonal of the occupied states
+        # where we still know some information from the partially converged
+        # nonoccupied states (in particular, they are Rayleigh-Ritz wrt to H).
+        # /!\ this is only implemented for insulators at the moment, WIP for
+        # /!\ metals (the old sternheimer is still used, see the "else" case)
+
+        # computed but nonconverged bands
+        Y = ψk_extra
+
+        # We put things into the form δψn = Yαn + Z ∈ Ran(Q)
+        # where Y = ψk_extra and Z ∈ Ran(R),
+        # R being the projector onto the orthogonal of computed states.
+        # This can be summerized as the following:
+        #
+        # <--- P ----><---- Q -----------
+        #                      <--- R ---
+        # |----------|--------|----------
+        # 1          N     N + N_extra
+        # <-- cvg --->
+        # <---- computed ----->
+
+        # projector onto the orthogonal of the computed states
+        R(ϕ) = ϕ - ψk * (ψk' * ϕ) - Y * (Y' * ϕ)
+
+        # Y are not converged but have been Rayleigh-Ritzed
+        # so YHY should be a real diagonal matrix
+        H(ϕ) = Hk * ϕ - εnk * ϕ
+        YHY = real.(Diagonal(εk_extra .- εnk))
+        YHYi = real.(Diagonal(1 ./ (εk_extra .- εnk)))
+
+        # computing products once and for all
+        HY = H(Y)
+        HYIY = HY * YHYi * Y'
+
+        b = -Q(rhs)
+
+        # 1) solve for Z
+        # --------------
+        # writing αn as a function of Z, we get that Z solves the system (in
+        # Ran(R))
+        #
+        # R * (1 - HYIY) * (H-εn) * R * Z = R * (1 - HYIY) * b
+        #
+        bb = R(b -  HYIY * b)
+        function RAR(ϕ)
+            HRϕ = H(R(ϕ))
+            ARϕ = HRϕ - HYIY * HRϕ
+            R(ARϕ)
+        end
+        precon = PreconditionerTPA(basis, kpoint)
+        precondprep!(precon, ψnk)
+        function R_ldiv!(x, y)
+            x .= R(precon \ R(y))
+        end
+        J = LinearMap{eltype(ψk)}(RAR, size(Hk, 1))
+        # tol_cg should not be too tight, and in particular not be
+        # too far below the error in the ψ. Otherwise Q and H
+        # don't commute enough, and an oversolving of the linear
+        # system can lead to spurious solutions
```

Comment:
Would be interesting to explore this a bit theoretically. R should help with the non-commutation, right? So the condition is something like tol >> ||R [Q,H] R||? Can we prove that something like this holds? Doe the gap play a role? If so which one? Generally speaking I haven't seen any error analysis for response-type linear problems, it'd be interesting to make even a rough analysis.

Reply:
Yes, that is exactly what we intend to explore with @mfherbst after I implement your review ! From what we already saw, R does help indeed.